[MRG] Add "grouped" option to Scaler classes

stephen-hoover opened the pull request:
As per discussion in issue #4892 , add a "per_feature" option to the Scaler classes. When False (defaults to the previous behavior of True), this modifies behavior such that scaling is based on a consideration of the entire data array at once, instead of one feature at a time. Also allow "axis=None" in addition to axis=0 or axis=1 in the standalone scaling functions.

This PR includes tweaks to functions in the "sparsefuncs" module where it makes axis=None behavior easier to code.

TODO:
- [x] Add "per_feature" option to RobustScaler
- [x] Add "axis=None" option to `preprocessing.data.scale`
- [x] Add "axis=None" option to `preprocessing.data.maxabs_scale`
- [x] Add "axis=None" option to `preprocessing.data.robust_scale`

stephen-hoover committed:
ENH Add "per_feature" option to Scalar classes

Defaults to True, which is the previous behavior. Also allow the scalar functions to take "axis=None" inputs in addition to 0 or 1.
Includes some tweaks to the sparsefuncs helper functions to deal with `axis=None` inputs.
Includes a new function `sparse_mean_variance` in "sparsefuncs_fast.pyx" to find the variance of a sparse array; that didn't seem to exist yet. This is one line in pure Python, but writing in Cython is faster and avoids extra memory use (as long as the input array has dtype np.float32 or np.float64).

stephen-hoover commented:
Tagging @amueller because this PR is a product of the SciPy 2015 sprints. I think this is ready for review now.

TomDLT commented on sklearn/utils/sparsefuncs.py:
```diff
@@ -54,31 +55,34 @@ def inplace_csr_row_scale(X, scale):
-    variances: float array with shape (n_features,)
-        Feature-wise variances
+    variances: float array with shape (n_features,) or scalar
+        Axis variances (or array mean if `axis` is None)
```
Axis variances (or array variance if `axis` is None)

TomDLT commented on sklearn/utils/tests/test_sparsefuncs.py:
```diff
@@ -95,6 +95,41 @@ def test_mean_variance_axis1():
+    X[2, 1] = 0
+    X[4, 3] = 0
+    X_lil = sp.lil_matrix(X)
+    X_lil[1, 0] = 0
```
```python
X[1, 0] = 0
X_lil = sp.lil_matrix(X)
```
?

ogrisel replied:
I think this was intentional to check that the code is actually dealing with "stored" zeros as opposed to zeros that are implicitly represented by non-stored entries.

TomDLT replied:
Ok it makes sense

TomDLT commented on sklearn/utils/tests/test_sparsefuncs.py:
```diff
@@ -95,6 +95,41 @@ def test_mean_variance_axis1():
+
+    X = X.astype(np.float32)
+    X_csr = X_csr.astype(np.float32)
+    X_csc = X_csr.astype(np.float32)
```
X_csc = X_csc.astype(np.float32)

TomDLT commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -227,6 +245,40 @@ def test_min_max_scaler_iris():
+    assert_array_almost_equal(X, X_trans_inv)
+
+    # not default params: min=1, max=2
+    scaler = MinMaxScaler(feature_range=(1, 2))
```
copy-paste of test_min_max_scaler_iris().
You need to add `per_feature=False` and change the assertions accordingly
:)

TomDLT commented:
Look good !

stephen-hoover committed:
TST: Fix copy-paste errors in scaler and sparsefuncs tests

stephen-hoover commented:
@TomDLT , thank you for the careful read! I fixed the issues you pointed out. I left it in a separate commit for now so you could check only my fixes; if you're satisfied, I'll squash them into the previous commit to keep the history clean.

On the `test_mean_variance_axisnone`, I should have read a bit more carefully -- I copy-pasted directly from the `test_mean_variance_axis0` and `axis1` functions, and those errors were present there. Rather than have the same tests copy-pasted six times, I combined all three of the axis tests into one function.

stephen-hoover committed:
BUG: Fix `sparse_mean_variance` for non-double/floats

Was inadvertently returning 0 instead of the correct answer for datatypes other than np.float32 and np.float64.

stephen-hoover commented:
...and another new commit to fix a nasty bug in the `sparsefuncs_fast.sparse_mean_variance` function I wrote. Includes test case which failed in the presence of the bug.

ogrisel commented:
The tests fail under windows with:

```
======================================================================
ERROR: sklearn.utils.tests.test_sparsefuncs.test_mean_variance_all_axes
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Python34-x64\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "C:\Python34-x64\lib\site-packages\sklearn\utils\tests\test_sparsefuncs.py", line 47, in test_mean_variance_all_axes
    for dtype in [np.float32, np.float64, np.float128]:
AttributeError: 'module' object has no attribute 'float128'
```

Apparently the availability of `numpy.float128` and its actual precision is platform dependent:

http://stackoverflow.com/questions/9062562/what-is-the-internal-precision-of-numpy-float128

Please adjust the test to run the test with `np.float128` only if `hasattr(np, 'float128')`.

ogrisel commented:
More details in http://stackoverflow.com/a/17023995/163740 . Based on this information I think we should just not use `np.float128` and instead stick to `np.float32` and `np.float64`. I don't think we need more than 64 bit precision for floating point values in a machine learning context anyway.

stephen-hoover commented:
As per that stackoverflow post, would np.longdouble be okay? I agree that float32 and float64 are the only ones that really matter. I just want to be certain that `sparsefuncs_fast.sparse_mean_variance` works properly for all input types. It has different code for three possible input types: np.float32, np.float64, and everything else. I tried testing with float16 first, but that would require adjusting the allowed tolerance on the `assert_array_almost_equal` checks. Using something with higher precision is a simpler solution, if it works.

ogrisel commented:
> As per that stackoverflow post, would np.longdouble be okay?

I think we don't need to test that. This is a YAGNI. Let's keep the tests simple to maintain and focus on the useful things.

ogrisel commented:
> It has different code for three possible input types: np.float32, np.float64, and everything else. I tried testing with float16 first, but that would require adjusting the allowed tolerance on the assert_array_almost_equal checks. Using something with higher precision is a simpler solution, if it works.

`np.float16` might be useful for some problems with tight memory constraints. +1 for adapting the tests for this. Otherwise you can also try with a complex dtype although I don't have a usecase in a machine learning context.

stephen-hoover commented:
@ogrisel , done. The test now uses `np.float16` in addition to `np.float32` and `np.float64`. The `lil_matrix` had trouble with `np.float16`, so I switched to `coo_matrix` for the invalid-sparse-style test.

ogrisel commented on sklearn/utils/sparsefuncs_fast.pyx:
```diff
@@ -48,6 +50,63 @@ def csr_row_norms(X):
 @cython.cdivision(True)
+def sparse_mean_variance(X):
+    """Compute the variance of a sparse matrix.
+    Will copy data to an array of doubles if input data
```
Please insert a blank line after the first line of the docstring.

ogrisel commented on sklearn/utils/sparsefuncs_fast.pyx:
```diff
@@ -48,6 +50,63 @@ def csr_row_norms(X):
+        for i_datum in xrange(len_data):
+            datum = X_data_f[i_datum]
+            sum += datum
+            sum_squares += datum * datum
```
I think this function should be rewritten to leverage Cython fused types:

```cython
from cython cimport floating
from numpy import float16_t, float32_t, float64_t, ndarray

ctypedef fused myfloating:
    float16_t
    float32_t
    float64_t

def sparse_mean_variance(X):
    # check data type and make a copy with astype if necessary
    # then call _floating_mean_variance

cdef _floating_mean_variance(ndarray[floating, ndim=2] array):
    # put the actual computation here

```

This would also allow to avoid a memory copy with float16 data. To check the case where a memory copy is required, just use int32 or int64 typed input.

ogrisel replied:
Actually is it really worth introducing a new Cython function just for that? Why not use `X.data.mean()` and `X.data.var()` directly?

stephen-hoover replied:
The `float16` is not supported in C, so there's no way to include it in compile Cython code. The best reference I found Googling on that is at https://groups.google.com/forum/#!topic/cython-users/a_gRQJzauJ4 . That thread is old now, but my current version of Cython also does not define a float16 type.

`X.data.mean()` and `X.data.var()` aren't the same as `X.mean()` and `X.var()`, since they ignore the implicit zeros in the sparse array. The scipy sparse matrices support `X.mean()`, but for some reason not `X.var()`. I can think of two ways to do this from pure Python:
`variance = (arr.data ** 2).sum() / (arr.shape[0] * arr.shape[1]) - arr.mean() ** 2`
or 
```
def meanvar_sparse(arr):
    data_mean = arr.data.mean()
    data_var = arr.data.var()
    ndata = len(arr.data)
    narr = arr.shape[0] * arr.shape[1]
    data_sum = data_mean * ndata
    data_sum_squares = (data_var + (data_mean * data_mean)) * ndata
    
    arr_mean = (data_sum / narr)
    arr_var = data_sum_squares / narr - arr_mean ** 2
    
    return arr_mean, var
```

Both are slower than the Cython code, and the first uses extra memory by creating a new array of length `len(X.data)` as an intermediate result. Since I'd need a new function anyway, and the scikit-learn code already included Cython functions to compute means and variances along columns and rows of sparse matrices, it seemed reasonable to add another Cython function for the entire array variance. 

I actually tried fused types before I submitted this PR. It does make the code simpler, but I was surprised to find that it was slower. Since there's not a large number of cases to support (float32 and float64 seem like the only ones that matter), I opted for the extra explicit case rather than fused types.

Following are some results of speed testing with `%timeit`. I'm comparing the current state of this PR (calling `sparsefuncs.mean_variance_axis(arr, axis=None)`) to a Cython function `_floating_mean_variance` defined for fused float32 and float64 types (fully typed, so that there's no Python lines left in the annotate), to the Python `meanvar_sparse` defined above. My test data are randomly-generated `csr_matrix` objects with either shape 5kx5k and 100k non-zero entries, or 50k x 50k and 1M or 10M non-zero entries. Results:

Non-zero entries | No fused types | Fused types | Python
-----------------------|----------------------|------------------|----------
100,000 | 112 us | 177 us | 325 us
1,000,000 | 1.21 ms | 1.36 ms | 5.25 ms
10,000,000 | 11.6 ms | 11.8 ms | 73 ms

If I switch the `_floating_mean_variance` function to use np.float64_t instead of a fused type, its speed equals the current `sparse_mean_variance`, which confirms that the extra time is overhead in detecting the type and choosing which function to use. The constant overhead is not large, but it still didn't seem worth using the fused types here.

Given this, my preference is to remain with the current `sparse_mean_variance` function.

ogrisel commented on sklearn/utils/sparsefuncs_fast.pyx:
```diff
@@ -48,6 +50,63 @@ def csr_row_norms(X):
 @cython.boundscheck(False)
 @cython.wraparound(False)
 @cython.cdivision(True)
```
I think you can put that those once and for all at the top of the `.pyx` file as:

```cython
# cython: cdivision=True
# cython: boundscheck=False
# cython: wraparound=False
```

stephen-hoover committed:
MAINT Cython settings to module level in sparsefuncs_fast

Every function in the module used `@cython.boundscheck(False)`, `@cython.wraparound(False)`, and `@cython.cdivision(True)`. Move those settings to module level to simplify the code.

stephen-hoover commented:
Pinging the people who reviewed this PR earlier: @TomDLT , @ogrisel . I think I've addressed all comments. I rebased on the current master, and tests are passing. There's two commits in here that I would squash after I get a sign-off.

TomDLT commented on sklearn/utils/sparsefuncs.py:
```diff
@@ -54,31 +55,34 @@ def inplace_csr_row_scale(X, scale):
-    if axis not in (0, 1):
+    if axis not in (0, 1, None):
         raise ValueError(
             "Unknown axis value: %d. Use 0 for rows, or 1 for columns" % axis)
```
should we update this message?

TomDLT commented:
I don't know if not handling `np.float16` is a problem or not.

Otherwise it looks good to me.

stephen-hoover committed:
MAINT Update error message in `mean_variance_axis`

stephen-hoover commented:
This code will handle `np.float16`, it'll just have to copy the array contents to an `np.float64` array first. I don't think there's a way around a copy. `np.float16` is non-standard, and Cython doesn't handle it.

I changed the error message you pointed out.

cmarmo commented:
Hi @stephen-hoover ... sorry, apparently this pull request got lost.... are you still interested in finishing it? Thanks for your patience.

stephen-hoover commented:
@cmarmo - Oh, wow, I'd forgotten about this PR. I will take a look at it and try to figure out how much work it would take to bring it up to date. I don't have a huge amount of free time anymore, so I couldn't redo it from scratch. Is this feature still useful? If you're asking, I'm guessing the answer is yes.

jnothman commented:
I don't think it would hurt, but I also don't think there's substantial demand for this change

cmarmo commented:
@stephen-hoover , thanks for your answer. I'm checking to relabel if necessary, in particular because your PR was "Waiting for Reviewer". If you are not available anymore to work on this, there is no problem.

thomasjpfan commented:
Since the original issue is closed, I am closing this PR as well. Thank you for working on this PR.
